### ISSUE TITLE
custom configuration for global timeline creation

### ISSUE THREAD

**Issue by ftanuma** [amzn/hawktracer#56]
Hi,

ht_global_timeline_get() makes it easy to use different timeline per thread, but we cannot use different configuration to create it. For example, we cannot pass larger(or smaller) buffer size.

Also, it makes a bit tricky to set up a same listener for all the global timelines. For example, if a threads are created in modules that doesn't have central manager that knows those threads, one way to set same listener to those timeline is to make global access to the listener and set it to each timeline. But it makes timeline setup codes here and there.

One idea to solve this is to make an API that we can pass a custom creation/setup function to hawktracer, which will be called back from ht_global_timeline_get(), if it is set.

How do you think?
Thanks!

**Comment by loganek** [comment]
Hi @ftanuma , thanks for your report.
Regarding global timeline's configuration - it's not implemented but that was one of the purpose of having `argc` and `argv` parameters in `ht_init` method - would it be sufficient for you to pass global timeline parameters through that function? It'd be something like:
```cpp
const char* argv = {"--ht-global-timeline-buffer-size", "4096"};
ht_init(2, argv);
```

For setting up the same listener for all global timelines, you only need to set it up for one timeline (i.e. in one thread) and global timelines in other threads will re-use the same listener, so:
```cpp
void foo()
{
  ht_timeline_register_listener(ht_global_timeline_get(), listener_callback, NULL);
  std::thread([] {
    HT_TIMELINE_PUSH_EVENT(ht_global_timeline_get(), EventType, some_parameter); // <- this should already have listener registered
}
```
In other words, all the global timelines share the same listeners, and if that doesn't work, this is a bug, so please let me know and I can fix that.

**Comment by ftanuma** [comment]
Hi @loganek , sorry for late response.

Regarding the configuration, yes argc/argv would work.

Regarding the listener, you are right. I was manually creating each timeline by ht_timeline_create() and I didn't notice that the 4th parameter(const char* listeners) is the key for same listener. So I was passing different names.
Now if I specify same const char*, it uses same listener for subsequently created timelines.

Thanks!

**Comment by loganek** [comment]
Hi @ftanuma 
I just merged the change. You can set buffer size of the global timeline by calling `ht_init` with command line special argument:

```c
const char* args[] = {"your-app", "--ht-global-timeline-buffer-size", "4096"};
ht_init(sizeof(args)/sizeof(args[0]), (char**)argv);
```
Feel free to re-open the ticket if there's anything else you need for this feature, or open a new issue if you have more feature requests.

**Comment by ftanuma** [comment]
Hi @loganek,

Thank you for implementing this. This made our implementation a lot simpler!

Btw, I haven't tested but there might be issue with ht_init(). ht_init() and ht_deinit() seems to allow us to call multiple times. What happens if I do,

`const char* args[] = {"your-app", "--ht-global-timeline-buffer-size", "0"};
ht_init(sizeof(args)/sizeof(args[0]), (char**)args);`

then some other module in the system do this?
`ht_init(0, nullptr);`

Do we still get global timeline with buffer size 0, or something default?

**Comment by loganek** [comment]
@ftanuma 
buffer size is set only if `--ht-global-timeline-buffer-size` flag is specified, so if your first call of `ht_init` specifies buffer size parameter (e.g. buffer_size=X), and the second call does not, the buffer size remains unchanged (i.e. it will still be X).  
One thing to keep in mind is that this setting only affects new instances of global timeline, so if you already have an instance of global timeline in any of the threads before calling `ht_init` with `buffer-size` parameter, those instances will have the old value of buffer size (e.g. the default one).
In general, I highly recommend to have a single call of `ht_init` in the program; multiple `ht_init` calls was added only to support some extreme cases, but in 99% of cases you don't need (and probably don't want) to do it :)

**Comment by ftanuma** [comment]
@loganek Thank you for clarification.
